Description: Nova patches for Project Calico
 Please see https://github.com/Metaswitch/calico-nova/tree/ubuntu_2014.1 for
 details.

---

--- nova-2014.1.orig/nova/network/model.py
+++ nova-2014.1/nova/network/model.py
@@ -38,6 +38,7 @@ VIF_TYPE_802_QBH = '802.1qbh'
 VIF_TYPE_MLNX_DIRECT = 'mlnx_direct'
 VIF_TYPE_MIDONET = 'midonet'
 VIF_TYPE_OTHER = 'other'
+VIF_TYPE_ROUTED = 'routed'
 
 # Constants for dictionary keys in the 'vif_details' field in the VIF
 # class
--- nova-2014.1.orig/nova/virt/libvirt/vif.py
+++ nova-2014.1/nova/virt/libvirt/vif.py
@@ -314,6 +314,17 @@ class LibvirtGenericVIFDriver(LibvirtBas
 
         return conf
 
+    def get_config_routed(self, instance, vif, image_meta,
+                           inst_type):
+        conf = super(LibvirtGenericVIFDriver,
+                     self).get_config(instance, vif,
+                                      image_meta, inst_type)
+
+        dev = self.get_vif_devname(vif)
+        designer.set_vif_host_backend_ethernet_config(conf, dev)
+
+        return conf
+
     def get_config_mlnx_direct(self, instance, vif, image_meta,
                                inst_type):
         conf = super(LibvirtGenericVIFDriver,
@@ -379,6 +390,11 @@ class LibvirtGenericVIFDriver(LibvirtBas
                                            vif,
                                            image_meta,
                                            inst_type)
+        elif vif_type == network_model.VIF_TYPE_ROUTED:
+            return self.get_config_routed(instance,
+                                           vif,
+                                           image_meta,
+                                           inst_type)
         else:
             raise exception.NovaException(
                 _("Unexpected vif_type=%s") % vif_type)
@@ -583,6 +599,14 @@ class LibvirtGenericVIFDriver(LibvirtBas
         except processutils.ProcessExecutionError:
             LOG.exception(_("Failed while plugging vif"), instance=instance)
 
+    def plug_routed(self, instance, vif):
+        """Plug a routed virtual interface
+        """
+        super(LibvirtGenericVIFDriver,
+              self).plug(instance, vif)
+        dev = self.get_vif_devname(vif)
+        linux_net.create_tap_dev(dev)
+
     def plug(self, instance, vif):
         vif_type = vif['type']
 
@@ -611,6 +635,8 @@ class LibvirtGenericVIFDriver(LibvirtBas
             self.plug_mlnx_direct(instance, vif)
         elif vif_type == network_model.VIF_TYPE_MIDONET:
             self.plug_midonet(instance, vif)
+        elif vif_type == network_model.VIF_TYPE_ROUTED:
+            self.plug_routed(instance, vif)
         else:
             raise exception.NovaException(
                 _("Unexpected vif_type=%s") % vif_type)
@@ -744,6 +770,17 @@ class LibvirtGenericVIFDriver(LibvirtBas
         except processutils.ProcessExecutionError:
             LOG.exception(_("Failed while unplugging vif"), instance=instance)
 
+    def unplug_routed(self, instance, vif):
+        """Unplug a routed virtual interface.
+        """
+        super(LibvirtGenericVIFDriver,
+              self).unplug(instance, vif)
+        dev = self.get_vif_devname(vif)
+        try:
+            linux_net.delete_net_dev(dev)
+        except processutils.ProcessExecutionError:
+            LOG.exception(_("Failed while unplugging vif"), instance=instance)
+
     def unplug_iovisor(self, instance, vif):
         """Unplug using PLUMgrid IO Visor Driver
 
@@ -793,6 +830,8 @@ class LibvirtGenericVIFDriver(LibvirtBas
             self.unplug_mlnx_direct(instance, vif)
         elif vif_type == network_model.VIF_TYPE_MIDONET:
             self.unplug_midonet(instance, vif)
+        elif vif_type == network_model.VIF_TYPE_ROUTED:
+            self.unplug_routed(instance, vif)
         else:
             raise exception.NovaException(
                 _("Unexpected vif_type=%s") % vif_type)
